File: JAVA/350 - Intersection of Two Arrays II.java
/*
 350. Intersection of Two Arrays II (E)

 Approach 1 : Sort + compare one by one
 Time: O(4n), two sorts + compare + copy to array
 Space: O(result.size())

*/

/* Approach 1 */
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        if (nums1.length == 0 || nums2.length == 0) return new int[] {};
        List<Integer> list = new ArrayList<Integer>();
        int i = 0;
        int j = 0;
        
        // Sort two lists
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        
        // Compare one by one
        while (i < nums1.length && j < nums2.length) {
            if (nums1[i] == nums2[j]) {
                list.add(nums1[i]);
                i++;
                j++;
            }
                
            else if (nums1[i] < nums2[j]) {
                i++;
            }
            else {
                j++;
            }
        }
        
        // Copy list to array
        int k = 0;
        int[] res = new int[list.size()];
        for (int x : list) {
            res[k++] = x;
        }
        
        return res;
    }
}

File: C++/350-Intersection_of_Two_Arrays_II.cpp
/*
 350. Intersection of Two Arrays II
 Easy
*/

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());
        vector<int>::iterator it1 = nums1.begin(), it2 = nums2.begin();
        vector<int> result;
        while (it1 != nums1.end() && it2 != nums2.end())
        {
            if (*it1 > *it2)
                it2++;
            else if (*it1 < *it2)
                it1++;
            else
            {
                result.push_back(*it1);
                it1++;
                it2++;
            }
        }
        return result;
    }
};
